plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/WebProjectUtil.java

```diff
@@ -45,22 +53,63 @@
 
   static final String WEB_INF = "WEB-INF";
 
+  /**
+   * Create a file in the project's WEB-INF, respecting the order and tags on the WTP virtual
+   * component model. Specifically, we create files and directories in the
+   * {@code &lt;wb-resource&gt;} with the {@code defaultRootSource} tag, if found. We do not
+   * overwrite the file.
+   * 
+   * @param project the hosting project
+   * @param filePath the path of the file within the project's {@code WEB-INF}
+   * @param contents the content for the file
+   * @see <a href="https://bugs.eclipse.org/bugs/show_bug.cgi?id=448544">Eclipse bug 448544</a>
+   */
+  public static IFile createWebInfFile(IProject project, IPath filePath, InputStream contents,
```

Comment:
I gave some thoughts on this, and I think this isn't perfect yet. About the reason why we go through all the hassle of respecting the WTP virtual component model, you said earlier that

> ... there *may* be multiple directories where it (`appengine-web.xml`) could exist. .... You *could* have a `WebContent`, `web`, `src/main/webapp` all in the same project.
>
> In this case we're only creating the `appengine-web.xml` *if* it doesn't already exist.

However, it will still create `appengine-web.xml` in this case if the file isn't at the `defaultRootSource`. For example,

1. `src/main/webapp` is tagged as `defaultRootSource`.
1. `src/main/webapp/WEB-INF/appengine-web.xml` doesn't exist.
1. `WebContent` is another web-resource source-path.
1. `WebContent/WEB-INF/appengine-web.xml` exists.

Then you'll create a brand new `appengine-web.xml`, even though the project already had its own file.

But I'd say at least we are not making matters worse from now, so let's create a low priority bug rather than making this PR more complicated. Having multiple web resource folders sounds like a corner case.

Reply:
> Then you'll create a brand new appengine-web.xml, even though the project already had its own file.

The behaviour you're describing is deliberate: I'm assuming that the caller will use `findInWebInf()` first and only call `createWebInfFile()` if the file is not found.  Since `createWebInfFile()` is provided the file-contents, it didn't seem helpful to search the various `<wb-resource>` directories again.  And with m2e's content-filtering, it might be useful to be able to explicitly create a file that m2e might then re-filter.

I'll change the javadoc for `createWebInfFile()` to make it explicit that resolving is done outside.